libcudacxx/include/cuda/std/detail/libcxx/include/__bit_reference

```diff
@@ -147,310 +174,122 @@ class __bit_const_reference
-  static const int __bits_per_word = _It::__bits_per_word;
-  // do first partial word
-  if (__first.__ctz_ != 0)
-  {
-    __storage_type __clz_f = static_cast<__storage_type>(__bits_per_word - __first.__ctz_);
-    __storage_type __dn    = _CUDA_VSTD::min(__clz_f, __n);
-    __storage_type __m     = (~__storage_type(0) << __first.__ctz_) & (~__storage_type(0) >> (__clz_f - __dn));
-    __storage_type __b     = *__first.__seg_ & __m;
-    if (__b)
-    {
-      return _It(__first.__seg_, static_cast<unsigned>(_CUDA_VSTD::__libcpp_ctz(__b)));
-    }
-    if (__n == __dn)
-    {
-      return __first + __n;
-    }
-    __n -= __dn;
-    ++__first.__seg_;
-  }
-  // do middle whole words
-  for (; __n >= __bits_per_word; ++__first.__seg_, __n -= __bits_per_word)
-  {
-    if (*__first.__seg_)
-    {
-      return _It(__first.__seg_, static_cast<unsigned>(_CUDA_VSTD::__libcpp_ctz(*__first.__seg_)));
-    }
-  }
-  // do last partial word
-  if (__n > 0)
-  {
-    __storage_type __m = ~__storage_type(0) >> (__bits_per_word - __n);
-    __storage_type __b = *__first.__seg_ & __m;
-    if (__b)
-    {
-      return _It(__first.__seg_, static_cast<unsigned>(_CUDA_VSTD::__libcpp_ctz(__b)));
-    }
-  }
-  return _It(__first.__seg_, static_cast<unsigned>(__n));
-}
+  using _It            = __bit_iterator<_Cp, false>;
+  using __storage_type = typename _It::__storage_type;
 
-template <class _Cp, bool _IsConst>
-__bit_iterator<_Cp, _IsConst> __find_bool_false(__bit_iterator<_Cp, _IsConst> __first, typename _Cp::size_type __n)
-{
-  typedef __bit_iterator<_Cp, _IsConst> _It;
-  typedef typename _It::__storage_type __storage_type;
   const int __bits_per_word = _It::__bits_per_word;
   // do first partial word
   if (__first.__ctz_ != 0)
   {
     __storage_type __clz_f = static_cast<__storage_type>(__bits_per_word - __first.__ctz_);
-    __storage_type __dn    = _CUDA_VSTD::min(__clz_f, __n);
+    __storage_type __dn    = (_CUDA_VSTD::min)(__clz_f, static_cast<__storage_type>(__n));
     __storage_type __m     = (~__storage_type(0) << __first.__ctz_) & (~__storage_type(0) >> (__clz_f - __dn));
-    __storage_type __b     = ~*__first.__seg_ & __m;
-    if (__b)
+    if (_FillVal)
     {
-      return _It(__first.__seg_, static_cast<unsigned>(_CUDA_VSTD::__libcpp_ctz(__b)));
+      *__first.__seg_ |= __m;
     }
-    if (__n == __dn)
+    else
     {
-      return __first + __n;
+      *__first.__seg_ &= ~__m;
     }
-    __n -= __dn;
+    __n -= __dn.__data;
     ++__first.__seg_;
   }
   // do middle whole words
-  for (; __n >= __bits_per_word; ++__first.__seg_, __n -= __bits_per_word)
-  {
-    __storage_type __b = ~*__first.__seg_;
-    if (__b)
-    {
-      return _It(__first.__seg_, static_cast<unsigned>(_CUDA_VSTD::__libcpp_ctz(__b)));
-    }
-  }
+  __storage_type __nw = __n / __bits_per_word;
+  _CUDA_VSTD::fill_n(_CUDA_VSTD::__to_address(__first.__seg_), __nw, _FillVal ? ~static_cast<__storage_type>(0) : 0);
```

Comment:
I believe that it would be cleaner if we would define a `constexpr auto _fill_val = _FillVal ? static_cast<__storage_type>(~0) : static_cast<__storage_type>(0);`

And then use that in the partial and full writes

Reply:
I am confused, because I think this is the only place this actual pattern shows up?